rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/TextDocumentState.java

```diff
@@ -49,75 +59,253 @@
+    // `CompletableFuture`) cannot be interrupted.
+
+    private final int period; // Milliseconds
+    private final Executor scheduler;
+
+    // At any point in time, only one delayed future to retry the request for
+    // the resource should be `scheduled`, tied with the total remaining delay.
+    // For bookkeeping, a *stamped reference* is used. The reference is the
+    // delayed future, while the stamp is the remaining delay *upon completion
+    // of the delayed future*.
+
+    private final AtomicStampedReference<@Nullable CompletableFuture<T>> scheduled;
+
+    // The underlying resource provider is represented abstractly in terms of
+    // two suppliers, each of which corresponds with a state of the underlying
+    // resource provider. `getIfInitialized` should return `null` iff the
+    // underlying resource provided is not-initialized.
+
+    private final Supplier<@Nullable CompletableFuture<T>> getIfInitialized;
+    private final Supplier<CompletableFuture<T>> initializeAndGet;
+
+    public Debouncer(Duration period,
+            Supplier<@Nullable CompletableFuture<T>> getIfInitialized,
+            Supplier<CompletableFuture<T>> initializeAndGet) {
+
+        this(Math.toIntExact(period.toMillis()), getIfInitialized, initializeAndGet);
+    }
+
+    public Debouncer(int period,
+            Supplier<@Nullable CompletableFuture<T>> getIfInitialized,
+            Supplier<CompletableFuture<T>> initializeAndGet) {
+
+        this.period = period;
+        this.scheduler = CompletableFuture.delayedExecutor(period, TimeUnit.MILLISECONDS);
+        this.scheduled = new AtomicStampedReference<>(null, 0);
+        this.getIfInitialized = getIfInitialized;
+        this.initializeAndGet = initializeAndGet;
+    }
+
+    public CompletableFuture<T> get(Duration delay) {
+        return get(Math.toIntExact(delay.toMillis()));
+    }
+
+    public CompletableFuture<T> get(int delay) {
+        return schedule(delay, false);
+    }
+
+    private CompletableFuture<T> schedule(int delay, boolean reschedule) {
+
+        // Get a consistent old stamp and old reference
+        var oldRef = scheduled.getReference();
+        var oldStamp = scheduled.getStamp();
+        while (!scheduled.weakCompareAndSet(oldRef, oldRef, oldStamp, oldStamp));
+
+        // Compute a new reference (delayed future to retry this method)
+        var delayArg = new CompletableFuture<Integer>();
+        var newRef = delayArg
+            .thenApplyAsync(this::reschedule, scheduler)
+            .thenCompose(Function.identity());
+
+        // Compute a new stamp
+        var delayRemaining = Math.max(oldStamp, delay);
+        var newStamp = delayRemaining - period;
+
+        // If the underlying resource provider is initialized, then return the
+        // future to get the resource
+        var future = getIfInitialized.get();
+        if (future != null && scheduled.weakCompareAndSet(oldRef, null, oldStamp, 0)) {
```

Comment:
From the documentation of `AtomicStampedReference.weakCompareAndSet`:
"May fail spuriously and does not provide ordering guarantees, so is only rarely an appropriate alternative to compareAndSet."
So why use the weak variant? The "May fail spuriously" part scares me ;-)

Reply:
Ah, good catch! I misremembered the docs. It should be "strong" `compareAndSet` indeed.